[P113] Add Region of Interest configuration

tonhuisman opened the pull request:
From a [Forum request](https://www.letscontrolit.com/forum/viewtopic.php?t=10214)

Features:
- Configuration options for the Region of Interest, allowing a smaller Field of View (but also reduced sensitivity & range)
- Updated documentation
- Add ROI selection via dragging an area on the Optical Center grid, or use click-lock for touch-devices

TODO:
- [x] Testing by requester(s)
- [x] Update documentation
- [x] Move javascript ~and css~ code to CDN (and WebStaticData.h)

tonhuisman committed:
[P113] Add Region of Interest configuration

thomastech commented:
Fantastic!

My validation results are summarized here:
https://www.letscontrolit.com/forum/viewtopic.php?p=70316#p70316

thomastech commented:
Any chance you can add a SPAD numbering chart to the documentation? This would help users with placing the SPAD array location.

Something like this one (screenshot from st.com user manual UM2555):

![spad_number_chart](https://github.com/user-attachments/assets/3baecb38-a329-45d7-bcb4-cba16a8f13f2)

**EDIT: NEVER MIND. I found that the numbering chart is already in your RTD file.**

TD-er commented:
Hmm I think we should let the user add some coordinate and 'radius' or square size or upper-left/lower-right coordinate.
These numbers are perfectly fine for internal use and with some coffee and/or sleep I guess you could come up with some simple formula to convert some coordinate to an internal number as there seems to be some logical ordering per quadrant.
So you only need to swap X and Y and invert X and Y for the bottom 2 quadrants and then the center is (0, 0)
Then only some offset is needed per quadrant and some step-size of 8 in the X-direction.

This can probably be done using some nifty bitshifting and bitwise OR's per quadrant :)

thomastech commented:
> Hmm I think we should let the user add some coordinate and 'radius'

I believe that is what the ROI plugin settings already allow us to do.

In my validation I entered a center SPAD location and the X/Y array sized to go around it.  For example, center=159, X=4 and Y = 4. My assumption is that the resulting sensor array pattern is what you see in the example below:

![SPAD_region1](https://github.com/user-attachments/assets/0baba86d-431e-4149-88ad-312a0db1e216)

However, if it is more complicated than that, we may need some extra settings in the plugin that make it easy to setup.

tonhuisman commented:
> For example, center=159, X=4 and Y = 4.

As 199 is the default physical center of the sensor, your image suggests that the Optical Center index should be 167.

thomastech commented:
Good catch. :)

gitjoska commented:
pretty nice implementation.
just a few questions/comments
After the update to this versions the values for both SPAD's are 4 and for ROI it is 0. Are those the STD-Values for the measurement which would meant y 4x4 spot on the right lower corner of the measurement area.

to define the center you cold also use x,y coordinates where 0,0 is the center and -6 (left and down) to +6 (up and right).

tonhuisman commented:
> After the update to this versions the values for both SPAD's are 4 and for ROI it is 0

Ah, good catch, I forgot to add a check for 0-values and reset to the original defaults (16x16 SPAD's and 199 for Optical Center). I'll fix that.


> to define the center you cold also use x,y coordinates where 0,0 is the center and -6 (left and down) to +6 (up and right).

Hmm, not sure yet, will think about how to implement that. I'm trying to avoid adding the quite regular table to the code, so will have to do something a bit more smart.

tonhuisman committed:
[Devices] Update/Add help links for Stats

TD-er commented:
> > Hmm I think we should let the user add some coordinate and 'radius'
> 
> I believe that is what the ROI plugin settings already allow us to do.
> 
> In my validation I entered a center SPAD location and the X/Y array sized to go around it. For example, center=159, X=4 and Y = 4. My assumption is that the resulting sensor array pattern is what you see in the example below:
> 
> ![SPAD_region1](https://private-user-images.githubusercontent.com/10354989/353653277-0baba86d-431e-4149-88ad-312a0db1e216.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI1NDI4ODUsIm5iZiI6MTcyMjU0MjU4NSwicGF0aCI6Ii8xMDM1NDk4OS8zNTM2NTMyNzctMGJhYmE4NmQtNDMxZS00MTQ5LTg4YWQtMzEyYTBkYjFlMjE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAxVDIwMDMwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFiMTMxNDI5NTVkMzIxYzk4ODg2YTgxMDA5MjQ5NDE5ODgyNDkxNWM3ZDBkOThjYzc1Y2JjNjAxNmRkMmNiOWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OBHWinV0YVgW7SW-CFXf8T7swZLOK8PIF-1wOGnjzc4)
> 
> However, if it is more complicated than that, we may need some extra settings in the plugin that make it easy to setup.

Just a quick write-up...
Assume internal coordinate system is (0, 0) for top left.
```c++
uint8_t getSPAD(uint8_t x, uint8_t y) {
  const uint8_t offset = ((x & 0x0F) << 3) + (y & 0x07);
  return (y > 7) ? 127 - offset : 128 + offset;
}
```

tonhuisman committed:
[P113] Check invalid ROI settings after in-place update

thomastech commented:
BTW, the library will force the center to 199 if the X or Y SPAD size exceeds 10. This is to ensure that they all fit in the sensor array. 

I think it would be helpful if the plugin also did this so the user can see the actual configuration after they submit their settings.

TD-er commented:
Ton told me in the chat, he was already fantasizing about JavaScript to select in a grid....
Well I don't really speak JS, but sounds extremely fancy.

I was just imaging to have some Paint-like experience where you select a brush and then 'point' at a spot.
So pre-defined set of grid-sizes and just point-and-shoot.

But that idea was like sooo 5-minutes ago. I think Ton is already firing up his editor and starts coding... or maybe going to sleep on it for one more night :)

tonhuisman commented:
> it would be helpful if the plugin also did this

That's included in the last commit I did 😉

tonhuisman committed:
Merge branch 'mega' into feature/P113-add-region-of-interest-settings

thomastech commented:
I've run into an issue that I didn't notice on the previous version; Currently testing to collect more information about the problem. 

**Issue:** After several hours of operation the VL53L1X sensor data does not change. Software (warm) rebooting does not fix it, but the VL53L1X plugin is disabled after the reboot. Restoring operation requires a power cycle. 

So far, slow I2C speeds (35KHz) don't help. And it's not an I2C bus lockup because there is a temperature sensor on the bus that continues to work correctly. 

I looked at the original VL53L1X source code and noticed that it ignores data read failures. It would be grand if it reported when data trouble has occurred. Maybe set all data values to zero? Or temporarily disable the plugin after several consecutive failures?

I have two ESP32 test devices with the same hardware configuration and both have experienced the problem. This morning I reverted one of them to the previous version (TD-er's last available test version without ROI). 

I will report back as soon as I have some useful information. That is to say, I think it would be helpful to confirm if the ROI update is involved. That kind of information may identify where to look. 

- Thomas

tonhuisman commented:
Thanks for the report. And the investigation, very helpful!

I have planned to apply the same 'state-machine' treatment to the VL53L1X library, that TD-er did to the VL53L0X library, but I'm currently working on another (ESPEasy) project. That _might_ get this issue corrected.

thomastech commented:
What is worrisome is that the affected sensor is in a state that prevents it from communicating. Because the only solution is a cold power cycle. So something must have happened that temporarily lobotomized it.

Test Conditions: 
I have one ESP32 running the code release that TD-er published Aug 12 (last Collection-B test bin before ROI). It's been running for a couple hours without issues. But it needs an overnight test before we know anything.

The other ESP32  is running your AUG 13 ROI test release. I had it set to 6X6 x 199, which has experienced the problem. I've changed it to the default 16X16 x 199 and rebooted. Will see if that makes a difference.

TD-er commented:
Sounds like the I2C bus might be hanging.

Whenever this does happen again, can you please check if it is even possible to get some results from an I2C scan?
If this is not possible in this situation where the sensor stops, can you try to swap both I2C pins, save the settings, do some I2C scan, revert those swapped I2C pins and see if the sensor starts working again.

I'm sure there once was some option to let ESPEasy detect a hanging I2C bus and then apply just those steps as I described above, but I can't find it in the Tools->Advanced page.

TD-er commented:
I2C can be used on nearly any frequency, almost down to DC.
However this also makes it highly depending on the actual clock pulses and when some clock pulse is missed or some extra "pulse" is inserted due to noise, both the I2C master (the ESP) and the sensor may get out of sync.
So either one may be waiting for that last clock pulse causing the I2C bus to become unusable.

By swapping the SDA and SCL, you generate enough random pulse combinations which causes any hanging I2C device to complete its reply.

tonhuisman commented:
> there once was some option to let ESPEasy detect a hanging I2C bus

That's been disabled for IDF 5.x 🤔

TD-er commented:
Ah yes....

thomastech commented:
> Sounds like the I2C bus might be hanging.

There's an LM75A sensor on the I2C bus. It functions without issue, so I don't think the I2C bus is hanging.

> Whenever this does happen again, can you please check if it is even possible to get some results from an I2C scan?

Will do.

> If this is not possible in this situation where the sensor stops, can you try to swap both I2C pins, save the settings, do some I2C scan, revert those swapped I2C pins and see if the sensor starts working again.

Will do.

> I'm sure there once was some option to let ESPEasy detect a hanging I2C bus and then apply just those steps as I described above, but I can't find it in the Tools->Advanced page.

The "Try clear I2C bus when stuck" check box was selected, no joy. But I see that Ton said the feature is disabled. So that box is just a decoy. :)

thomastech commented:
Good news. The issue is not related to the VL53L1X's ROI update. Because the older non-ROI version has failed with the "frozen data" problem.

1. I ran I2C scan and the VL53L1X & LM75A I2C devices were reported. Didn't expect to see the VL53L1X!
2. I reversed SDA and SCL pins and did a few I2C scans. No devices found, of course.
3. Restored correct I2C pinout and a I2C scan found the devices again. But the VL53L1X data is still frozen.

============================
NEW OBSERVATIONS:

1. I noticed that when the problem occurs, and I visit the VL53L1X plugin, change something and save, the plugin becomes disabled; Refuses to be enabled until a cold power cycle.

2. I also found that when things are working correctly, and I swap the I2C pins (on hardware page), as expected both I2C plugins become disabled. But in contrast, when the frozen data occurs the plugin does not detect the problem until I change a setting to it or do a warm/soft reboot. After doing those things the plugin becomes disabled. 

I think these two observations are due to the VL53L1X's value reading function ignores data read failures. Or maybe this info could be a clue to something else?

============================

I've run both sets of duplicated hardware for weeks. The mystery is how I never noticed the issue until recently. However, along the way I have increased the rule complexity. Maybe I've done something nasty in there, will do a fresh rule review tonight.

Currently the sensor reads are on a 2 sec rule timer. I will try five seconds and see if that changes anything. Maybe the reads are taking too long and I2C is corrupting the sensor by the relatively frequent access. Just thinking out loud, grasping at straws.

TD-er commented:
OK, so the sensor actually is in some kind of undefined state.
Not sure if the sensor does have a reset function. (or reset pin, but that might require rewiring).

The task remaining disabled is due to the "Check I2C devices when enabled" on the Tools->Advanced page.
You could try unchecking this. However if the sensor is actually in such an undefined state, I doubt it will work with this unchecked and the task set to enabled.
It might even cause ESPEasy to appear hanging as it may take forever to perform communications to the sensor.

So what you can do is this:
- Make sure the settings are saved with the task "enabled" state unchecked.
- Uncheck "Check I2C devices when enabled" on the tools->Advanced page
- enter the command `taskenable,<tasknr>` on the command field on the tools page. (with the correct task nr of course)

With the commands `taskenable` and `taskdisable` you can toggle the "Enabled" checkbox of a task, without saving the settings.
So if your unit becomes unresponsive, you can just power cycle (or restart) and ESPEasy will boot with the stored settings.

tonhuisman commented:
> The "Try clear I2C bus when stuck" check box was selected, no joy. But I see that Ton said the feature is disabled. So that box is just a decoy. :)

No, not a decoy, but disabled in all ESP32 LittleFS builds (using IDF 5.x framework) for technical reasons, working as intended in many other builds (unless dictated by lack of space in the binary).

thomastech commented:
> Not sure if the sensor does have a reset function. (or reset pin, but that might require rewiring).

There is an enable pin on the VL53L1X that might act as a reset, I don't know. It is mostly intended to allow selecting the device if several are on the same I2C bus. However, I painted myself into a corner because the sensors are potted in silicone and too late to add another wire.

> The task remaining disabled is due to the "Check I2C devices when enabled" on the Tools->Advanced page.
You could try unchecking this. However if the sensor is actually in such an undefined state, I doubt it will work with this unchecked and the task set to enabled.

That is how I understood this feature works. And since the sensor is stuck in a frozen data state, I prefer the plugin to be temporarily disabled. Because I have rules that check for a disabled plugin and the LCD can report the error. But this does not work in all scenarios since the sensor "failure" is only detected on plugin initialization.

> No, not a decoy, but disabled in all ESP32 LittleFS builds (using IDF 5.x framework) for technical reasons, working as intended in many other builds (unless dictated by lack of space in the binary).

I'm using a SPIFFS build, ESP_Easy_mega_20240813_collection_B_ESP32_4M316k. So perhaps the feature is working? But since the manual I2C pin swapping trick did not help, and I2C scan continues to work during the sensor's failed state, this issue is unlikely related to a I2C bus problem.

===========================
New Info:

1. Ran with 5 second updates (was 2 secs) in my rule file. No joy, still frozen data after awhile.

2. Disabled the email Notify plugin in case the hourly email sending event was corrupting the device. No joy.

3. Did other random experiments with plugin settings and edits to the rule file. So far no joy.

TD-er commented:
So you have it glued in on one side.
But how about the other end of the cable?
Could you toggle the 3V3 ?

And if not running at 3V3, could that also be the cause of issues where the levels might not be optimal?

If all those options do not yield a solution, I'm afraid we might need to look into the code but of course as you may understand it makes way more sense to start looking at someone else to search for ideas to launch the most wild theories what might cause these issues you're experiencing. ;)

thomastech commented:
> Could you toggle the 3V3 ?

Not loving that workaround, but it is something I have considered. I would still need to know when to cycle the power and re-init the plugin. Currently the plugin ignores the "frozen data" state and happily reports stale data.

I prefer to find the cause to the madness and fix it. So I am slogging through and kicking rocks. Fingers crossed I find something that exposes the villain.

Next on the list is to disable rules. So I will kick that rock and see where it lands.

tonhuisman commented:
You reported:

> 1. I ran I2C scan and the VL53L1X & LM75A I2C devices were reported. Didn't expect to see the VL53L1X!

But also:

> 1. change something and save, the plugin becomes disabled; Refuses to be enabled until a cold power cycle.

This seems a bit weird to me, as essentially, the check at initialization of the plugin is very similar (though not exactly the same) as the check that's done 'on the fly' during the read-cycle (there it gets disabled after 10 failed reads, not immediately).

Disabling the `Check I2C devices when enabled:` setting will inhibit the check altogether, as we have seen that some I2C devices really don't like to be poked that often.
If this solves the frozen-sensor syndrome, we have a flag available in the Device[] object to permanently disable this check for the plugin.

thomastech commented:
> Disabling the Check I2C devices when enabled: setting will inhibit the check altogether, as we have seen that some I2C devices really don't like to be poked that often.

Thank you for the tip. I will perform some tests with this function disabled.

thomastech commented:
> This seems a bit weird to me, as essentially, the check at initialization of the plugin is very similar (though not exactly the same) as the check that's done 'on the fly' during the read-cycle (there it gets disabled after 10 failed reads, not immediately).

I agree. And it would be ideal if the plugin detected the read issue on the fly so it could be disabled. I have a rule that reports a warning when this occurs. This could also allow me to have some hardware that power cycled the stuck sensor (hate to do that, would rather fix the problem) if the plugin gets disabled.

But after reviewing the VL53L1X code, I found that the only place it can detect I2C problems is in the P113_data->begin() function. The sensor read function does not have visibility to I2C errors because the Sparkfun library does not return a failure state.

I'm still confused why the "stuck data" problem was not noticed until recently. I might have just been lucky or maybe there is more to the story. Still investigating.

thomastech commented:
UPDATE: My problem is probably NOT a firmware bug. Because I found an old discussion on the ST Community Forum where a VL53L1X user was plagued with similar symptoms. And the issue was eventually fixed with hardware!

Full details are here: https://community.st.com/t5/imaging-sensors/vl53l1x-sensor-with-uld-api-stops-sending-data-after-some-time/td-p/360451

============

TL;DR: 
To summarized what was discussed, the ST factory rep insisted the user's issue was a bus hang caused by noise on I2C. But the user confirmed that did not occur. BTW, bus wire length was only 3-inches and proper supply decoupling was provided. A couple dozen devices were tested and all randomly ended up in a no-Ack state after awhile.

Out of frustration, the affected hardware's I2C pullups were reduced from 10K to 1K ohm. The exaggerated value solved his problem.

============

So yesterday I setup a controlled experiment where one of my devices was modified and the other unmodified. In my case, I added 1.5K ohms across the existing 5K ohm I2C pullups, so effectively 865 ohms.

So far the frozen data issue has not appeared on the modified hardware. But my unmodified device has not failed either, which is mind blowing. That fellow periodically failed all week long, but now it decides to behave.

I also added a rule hack to the unmodified unit to encourage it to disable the VL53L1X plugin during normal runtime when the sensor stops communicating. This will allow another rule to report the issue on the LCD. 

The hack is just a 30 minute rule timer that performs a TaskDisable / TaskEnable sequence. But I can't confirm this will work until the sensor "fails." Still waiting for that to happen.

TD-er commented:
Hmm strange they do need such a high pull-up resistors.
Do they use some capacitor parallel to the SDA or SCK pin?

thomastech commented:
The VL53L1X board doesn't have caps on the I2C signals. What I am using is a Chinese copy of the Adafruit design, but in a smaller form factor:
https://learn.adafruit.com/assets/105870

If the heavy-handed pullups solve the problem, then perhaps this is a hint that this board and/or chip is intolerant to any noise, and requires perfect I2C signals. My I2C cables are over 1.5 meters long, so I suppose I deserve to be punished.

At the moment, neither test unit has failed yet. It is unbelievable that the unmodified hardware is being a good boy.

TD-er commented:
Maybe less atmospheric disturbances at the moment?

Last few days (weeks?) I've heard a lot of complaints from a neighbor who's a radio amateur regarding those disturbances making it impossible to make long distance communications.

thomastech commented:
Could be atmospheric disturbances. Or witchcraft. Both are equally suspect.

The two test devices have been running without any problems; So happy to see them still working this morning. But still baffled that the unmodified unit is running good now. 

This morning I noticed both test devices were loaded with an earlier firmware version without the VL53L01X ROI feature. So I reverted back to Ton's Aug-13 release (ROI test build), which is the one I was using when I reported the problem. Let's see if that matters.

TL;DR : Still playing the waiting game. Will report back as I learn more.

thomastech commented:
Frustration levels are at DEFCON 1.

The problem did not return after flashing with the VL53L1X plugin with ROI. So that is a relief. But I have yet to see the "frozen data" problem on either unit since the day I modified one of them with the new resistor value.

So yesterday I unmodified it, back to 5K ohm I2C terminations. Still have not seen the problem. WTH.

TL;DR: Don't really know if the revised termination resistor values fix this. Because the problem has gone into hibernation. Will continue to test.

TD-er commented:
I think you have a Ghost Particle detector.
You don't even need to create an underground facility and cool your tools down to sub mK temperatures.

Maybe we should rename the plugin?
Perhaps Cern would help us debug it then?

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P113-add-region-of-interest-settings

tonhuisman committed:
Merge branch 'mega' into feature/P113-add-region-of-interest-settings

tonhuisman committed:
Merge branch 'feature/P113-add-region-of-interest-settings' of https://github.com/tonhuisman/ESPEasy-1 into feature/P113-add-region-of-interest-settings

thomastech commented:
Nothing new to report. Both test devices running without the I2C resistor value change and working fine.

I plan to add the I2C "1K ohm" resistor mod at the end of this validation test. Then test again for several days to confirm the lower value does not cause any problems. None expected, but I prefer to idiot-proof my decisions.

Still scratching my head on what made the problem appear for a week, then go away. Random bugs are the work of the devil.

TD-er commented:
> Random bugs are the work of the devil.

He should not have write access here.
Let me check just to be sure...

tonhuisman committed:
[P113] Add UI selection of ROI area (single area selection only)

tonhuisman commented:
@thomastech (and @TD-er)
Added selection of the ROI area in the UI (2-way 😛)
Please test and give feedback where appropriate, TIA 👍 

Teaser:
![p113_ROIselection2](https://github.com/user-attachments/assets/0c5b787e-744f-44db-8df5-f0629508ea44)

Also supported in dark mode UI:
![p113_ROIselection_dark](https://github.com/user-attachments/assets/75e76749-8722-403f-b61f-675f4d494c89)

thomastech commented:
That is very cool. Do you happen to have a bin for collection_B_ESP32_4M316k?

tonhuisman commented:
> Do you happen to have a bin for collection_B_ESP32_4M316k?

That's somewhat hidden in the Actions tab, at the top of this page, where you can find all recent push-builds and mega builds when a PR is merged. And there is also the checklist at the bottom of this page ("All checks have passed") that links to the latest PR build output.
Latest (overview) for this PR is [here](https://github.com/letscontrolit/ESPEasy/actions/runs/10856485440)

thomastech commented:
Got it loaded and it is working. The ROI layout tool is very nice.

But noticed a strange bug. Now at any reboot, one of my three switch plugins reports the incorrect switch state. I can clear it by pressing the switch for a few moments, then it acts OK.

Flashing with other bins does not cause this oddity. I haven't had time to look into it further. Will do that soon.

thomastech commented:
The Switch plugin bug is consistently repeatable. Appears after every reboot. Re-flashed my second device and it does it too. But only one of the switch plugins is affected, two others appear to be fine. But still debugging it.

TD-er commented:
Maybe also try switching an odd amount of times before reboot to see whether it was already an issue before or not.

thomastech commented:
The bug goes away if I uncheck the "Send Boot state" option on the affected Switch plugin. I'll use this as a workaround since I don't need to send the boot state.

To be clear, this bug only affects one of my switches, and only if I use the new ROI test build. Not seen in the past, appears to be a new problem.

> Maybe also try switching an odd amount of times before reboot to see whether it was already an issue before or not.

OK, I will try more builds from bygone days. But over the last couple months I've run about a dozen different bins on this device (Coffee Milk Level project) and have never noticed this particular bug before.

tonhuisman commented:
> this bug only affects one of my switches, and only if I use the new ROI test build.

Hmm, I'm _pretty_ sure I haven't touched the P001 code in this PR...

... but there are mods merged in from `mega`

thomastech commented:
> Hmm, I'm pretty sure I haven't touched the P001 code in this PR...
... but there are mods merged in from mega

Is there a build I can test that has all the latest merges, but without the ROI update to P113?

tonhuisman commented:
> Is there a build I can test that has all the latest merges, but without the ROI update to P113?

That would be from the last merge build, [over here](https://github.com/letscontrolit/ESPEasy/actions/runs/10516604256)

thomastech commented:
> That would be from the last merge build, [over here](https://github.com/letscontrolit/ESPEasy/actions/runs/10516604256)

That's one of the releases that I tried earlier today. It works fine.

TD-er commented:
And on the build which is failing, does it also fail when you don't use the P113 plugin? (e.g. not added to the task list and/or have it disabled)

thomastech commented:
I deleted the P113 plugin and the problem remains.

I discovered that during runtime, I can check and uncheck the Switch Plugin's "Send Boot state" option and the switch's state follows the check/uncheck selections. No reboot required. When in the "wrong" state, it can be corrected by operating the actual switch (or by unchecking the offending option).

There are three switches. Two of them (Door & Service) have this behavior. One (Jug) doesn't.

![Devices](https://github.com/user-attachments/assets/7a14570d-e93e-4d34-82be-5e3a443dd736)


Reflashing with a previous version eliminates this oddball bug. Moving back to the new test release (via OTA) and it returns.

I also performed a Tools->Factory_Reset using the function's default selections. Then reloaded a tar backup file (thanks Ton!) to restore everything. No improvement, bug still bugging me.

tonhuisman committed:
[Build] Possible fix for unstable builds

tonhuisman commented:
@thomastech TD-er found a possible issue for the differences in working for some Actions runs. A fix is applied and available from [this run](https://github.com/letscontrolit/ESPEasy/actions/runs/10870827625). Can you test this please?

TD-er commented:
If this isn't working, I still have some more 'tricks' up my sleeve to test.
I am also looking into the changed code related to P001_switch.

It seems there have been some changes related to 'send bootstate', when I merged the code duplication of P001, P009 and P019.
However I don't understand why this would make a difference between the builds you tested as this code is the same among both builds.

So let's see if this is an alignment issue which Ton tried to patch.

tonhuisman committed:
[Build] Possible fix, part 2, for unstable builds

tonhuisman commented:
Part 2 of the 'tricks up TD-er' sleeve', available from [this Actions run](https://github.com/letscontrolit/ESPEasy/actions/runs/10871718287)

thomastech commented:
Thanks for looking into the issue. I tried both builds, no joy.

Yesterday I also experienced some reboots when operating the DOOR switch. That is something that did not occur before.

thomastech commented:
Initially I only noticed one switch plugin was affected with this bug. But later I found two that did it. But further testing has shown that all three experience it.

When the default switch state is 1, the issue is essentially hidden. But if the default state is 0, the bug unexpectedly forces it to 1.

This all became clear to me after I arranged all switch logic states so they were in a 0 state during runtime. With that little change I can now confirm that all three plugins are affected.

TD-er commented:
Can you check this build when ready?
https://github.com/letscontrolit/ESPEasy/actions/runs/10873047163

N.B. it does not include the P113 plugin made by Ton.

thomastech commented:
> Can you check this build when ready?
https://github.com/letscontrolit/ESPEasy/actions/runs/10873047163

Tested. Working correctly. :)

> N.B. it does not include the P113 plugin made by Ton.

P113 is present and working, but appears to be an older version without the ROI layout feature.

This latest test build has an unusual behavior. When the /devices page is refreshed, all the data in the Values column appear with five decimal places. But after one second they change to the correct data precision.

TD-er commented:
> This latest test build has an unusual behavior. When the /devices page is refreshed, all the data in the Values column appear with five decimal places. But after one second they change to the correct data precision.

I know, working on it.... https://github.com/letscontrolit/ESPEasy/issues/5123

At least we now know what to fix for this (your) issue.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P113-add-region-of-interest-settings

tonhuisman committed:
[P113] Improve UI selection of ROI area for touch devices, code improvements

tonhuisman commented:
@thomastech I've rebased to include the SendBootstate fix in this build, and also added a click-lock feature to select the ROI area on touch-only devices, as click&select doesn't work there.
Build is being prepared in [this Actions run](https://github.com/letscontrolit/ESPEasy/actions/runs/10873769942).

tonhuisman committed:
[P113] Update documentation

tonhuisman committed:
[P113] Improve UI selection of ROI area for touch devices, code improvements

thomastech commented:
Latest test build is a winner. Problem resolved.

Despite the debugging detour, the new ROI selection tool is cool. Thanks for adding it.

TD-er commented:
Great!

Next build will be mainly about these kind of issues, so don't hesitate to report any of those "huh? that's odd" experiences.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P113-add-region-of-interest-settings

tonhuisman committed:
[P113] Further minify embedded js

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P113-add-region-of-interest-settings

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P113-add-region-of-interest-settings
